File: YASAudio/Basic/YASAudioData.m
//
//  YASAudioData.m
//  Copyright (c) 2015 Yuki Yasoshima.
//

#import "YASAudioData.h"
#import "YASAudioFormat.h"
#import "YASAudioUtility.h"
#import "YASMacros.h"
#import "NSException+YASAudio.h"

#if (!TARGET_OS_IPHONE && TARGET_OS_MAC)
#import "YASAudioChannelRoute.h"
#endif

typedef NS_ENUM(NSUInteger, YASAudioDataFreeType) {
    YASAudioDataFreeTypeNone,
    YASAudioDataFreeTypeFull,
    YASAudioDataFreeTypeWithoutData,
};

@interface YASAudioData ()
@property (nonatomic, strong) YASAudioFormat *format;
@end

@implementation YASAudioData {
    YASAudioDataFreeType _freeType;
    AudioBufferList *_audioBufferList;
}

- (instancetype)initWithFormat:(YASAudioFormat *)format frameCapacity:(UInt32)frameCapacity
{
    BOOL interleaved = format.isInterleaved;
    UInt32 bufferCount = interleaved ? 1 : format.channelCount;
    UInt32 stride = interleaved ? format.channelCount : 1;
    UInt32 bytesPerFrame = format.streamDescription->mBytesPerFrame;
    AudioBufferList *abl = YASAudioAllocateAudioBufferList(bufferCount, stride, frameCapacity * bytesPerFrame);

    return [self initWithFormat:format audioBufferList:abl needsFree:YES];
}

- (instancetype)initWithFormat:(YASAudioFormat *)format audioBufferList:(AudioBufferList *)abl needsFree:(BOOL)needsFree
{
    self = [super init];
    if (self) {
        if (!format || !abl) {
            YASRaiseWithReason(([NSString stringWithFormat:@"%s - Argument is nil. format(%@) audioBufferList(%p)",
                                                           __PRETTY_FUNCTION__, format, abl]));
            YASRelease(self);
            return nil;
        }

        _freeType = needsFree ? YASAudioDataFreeTypeFull : YASAudioDataFreeTypeNone;
        _audioBufferList = abl;
        self.format = format;

        UInt32 bytesPerFrame = format.streamDescription->mBytesPerFrame;
        UInt32 frameCapacity = abl->mBuffers[0].mDataByteSize / bytesPerFrame;
        _frameCapacity = frameCapacity;
        _frameLength = frameCapacity;
    }
    return self;
}

#if (!TARGET_OS_IPHONE && TARGET_OS_MAC)

- (instancetype)initWithFormat:(YASAudioFormat *)format
                          data:(YASAudioData *)data
           outputChannelRoutes:(NSArray *)channelRoutes
{
    return [self _initWithFormat:format data:data channelRoutes:channelRoutes isOutput:YES];
}

- (instancetype)initWithFormat:(YASAudioFormat *)format
                          data:(YASAudioData *)data
            inputChannelRoutes:(NSArray *)channelRoutes
{
    return [self _initWithFormat:format data:data channelRoutes:channelRoutes isOutput:NO];
}

- (instancetype)_initWithFormat:(YASAudioFormat *)format
                           data:(YASAudioData *)data
                  channelRoutes:(NSArray *)channelRoutes
                       isOutput:(BOOL)isOutput
{
    self = [super init];
    if (self) {
        if (!format || !data || !channelRoutes) {
            YASRaiseWithReason(
                ([NSString stringWithFormat:@"%s - Argument is nil. format(%@) data(%@) channelRoutes(%@)",
                                            __PRETTY_FUNCTION__, format, data, channelRoutes]));
            YASRelease(self);
            return nil;
        }

        if (format.channelCount != channelRoutes.count || format.isInterleaved) {
            YASRaiseWithReason(([NSString
                stringWithFormat:
                    @"%s - Invalid format. format.channelCount(%@) channelRoute.count(%@) isInterleaved(%@)",
                    __PRETTY_FUNCTION__, @(format.channelCount), @(channelRoutes.count), @(format.isInterleaved)]));
            YASRelease(self);
            return nil;
        }

        _freeType = YASAudioDataFreeTypeWithoutData;
        _audioBufferList = YASAudioAllocateAudioBufferListWithoutData(format.bufferCount, format.stride);
        self.format = format;

        const AudioBufferList *abl = data.audioBufferList;
        UInt32 bytesPerFrame = format.streamDescription->mBytesPerFrame;
        UInt32 frameCapacity = 0;

        for (UInt32 i = 0; i < format.channelCount; i++) {
            YASAudioChannelRoute *route = channelRoutes[i];
            UInt32 fromChannel = isOutput ? route.destinationChannel : route.sourceChannel;
            UInt32 toChannel = isOutput ? route.sourceChannel : route.destinationChannel;
            _audioBufferList->mBuffers[toChannel].mData = abl->mBuffers[fromChannel].mData;
            _audioBufferList->mBuffers[toChannel].mDataByteSize = abl->mBuffers[fromChannel].mDataByteSize;
            UInt32 frameLength = abl->mBuffers[0].mDataByteSize / bytesPerFrame;
            if (frameCapacity == 0) {
                frameCapacity = frameLength;
            } else if (frameCapacity != frameLength) {
                YASRaiseWithReason(
                    ([NSString stringWithFormat:@"%s - Invalid frame length. frameCapacity(%@) frameLength(%@)",
                                                __PRETTY_FUNCTION__, @(frameCapacity), @(frameLength)]));
            }
        }

        _frameCapacity = frameCapacity;
        _frameLength = frameCapacity;
    }
    return self;
}

#endif

- (id)copyWithZone:(NSZone *)zone
{
    YASAudioData *data = [[self.class allocWithZone:zone] initWithFormat:self.format frameCapacity:_frameCapacity];
    YASAudioCopyAudioBufferListDirectly(self.audioBufferList, data.mutableAudioBufferList);
    data.frameLength = _frameLength;
    return data;
}

- (void)dealloc
{
    if (_audioBufferList) {
        switch (_freeType) {
            case YASAudioDataFreeTypeFull:
                YASAudioRemoveAudioBufferList(_audioBufferList);
                break;
            case YASAudioDataFreeTypeWithoutData:
                YASAudioRemoveAudioBufferListWithoutData(_audioBufferList);
                break;
            default:
                break;
        }
        _audioBufferList = nil;
    }

    YASRelease(_format);

    _format = nil;

    YASSuperDealloc;
}

- (const AudioBufferList *)audioBufferList
{
    return _audioBufferList;
}

- (AudioBufferList *)mutableAudioBufferList
{
    return _audioBufferList;
}

- (void)setFrameLength:(UInt32)frameLength
{
    if (frameLength > _frameCapacity) {
        YASRaiseWithReason(
            ([NSString stringWithFormat:@"%s - Frame length is over capacity. frameLength(%@) frameCapacity(%@)",
                                        __PRETTY_FUNCTION__, @(frameLength), @(_frameCapacity)]));
        return;
    }

    if (_frameLength != frameLength) {
        _frameLength = frameLength;
    }
    UInt32 bytesPerFrame = self.format.streamDescription->mBytesPerFrame;
    UInt32 dataByteSize = frameLength * bytesPerFrame;
    YASAudioSetDataByteSizeToAudioBufferList(self.mutableAudioBufferList, dataByteSize);
}

- (UInt32)bufferCount
{
    return self.audioBufferList->mNumberBuffers;
}

- (UInt32)stride
{
    return self.audioBufferList->mBuffers[0].mNumberChannels;
}

- (YASAudioMutablePointer)pointerAtBuffer:(NSUInteger)buffer
{
    YASAudioMutablePointer pointer = {NULL};

    if (buffer >= self.audioBufferList->mNumberBuffers) {
        YASRaiseWithReason(
            ([NSString stringWithFormat:@"%s - Out of range. buffer(%@) abl.mNumberBuffers(%@)", __PRETTY_FUNCTION__,
                                        @(buffer), @(self.audioBufferList->mNumberBuffers)]));
    } else {
        pointer.v = self.audioBufferList->mBuffers[buffer].mData;
    }

    return pointer;
}

- (YASAudioMutablePointer)pointerAtChannel:(NSUInteger)channel
{
    YASAudioMutablePointer pointer = {NULL};

    const UInt32 stride = self.format.stride;

    if (stride > 1) {
        if (channel < self.audioBufferList->mBuffers[0].mNumberChannels) {
            pointer.v = self.audioBufferList->mBuffers[0].mData;
            if (channel > 0) {
                pointer.u8 += channel * self.format.sampleByteCount;
            }
        } else {
            YASRaiseWithReason(
                ([NSString stringWithFormat:@"%s - Out of range. channel(%@) mNumberChannels(%@)", __PRETTY_FUNCTION__,
                                            @(channel), @(self.audioBufferList->mBuffers[0].mNumberChannels)]));
        }
    } else {
        if (channel < self.audioBufferList->mNumberBuffers) {
            pointer.v = self.audioBufferList->mBuffers[channel].mData;
        } else {
            YASRaiseWithReason(
                ([NSString stringWithFormat:@"%s - Out of range. channel(%@) mNumberChannels(%@)", __PRETTY_FUNCTION__,
                                            @(channel), @(self.audioBufferList->mBuffers[0].mNumberChannels)]));
        }
    }

    return pointer;
}

- (void)clear
{
    self.frameLength = self.frameCapacity;
    YASAudioClearAudioBufferList(_audioBufferList);
}

- (void)clearWithStartFrame:(UInt32)frame length:(UInt32)length
{
    if ((frame + length) > self.frameLength) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Out of range. frame(%@) length(%@) frameLength(%@)",
                                                       __PRETTY_FUNCTION__, @(frame), @(length), @(self.frameLength)]));
        return;
    }

    UInt32 bytesPerFrame = self.format.streamDescription->mBytesPerFrame;

    for (UInt32 i = 0; i < self.bufferCount; i++) {
        Byte *data = self.audioBufferList->mBuffers[i].mData;
        memset(&data[frame * bytesPerFrame], 0, length * bytesPerFrame);
    }
}

- (BOOL)copyFromData:(YASAudioData *)fromData
{
    return [self copyFromData:fromData fromStartFrame:0 toStartFrame:0 length:fromData.frameLength];
}

- (BOOL)copyFromData:(YASAudioData *)fromData
      fromStartFrame:(UInt32)fromFrame
        toStartFrame:(UInt32)toFrame
              length:(UInt32)length
{
    if (!fromData) {
        YASRaiseWithReason(
            ([NSString stringWithFormat:@"%s - Argument is nil. fromData(%@)", __PRETTY_FUNCTION__, fromData]));
        return NO;
    }

    if (![fromData.format isEqualToAudioFormat:self.format]) {
        YASLog(@"%s - Format is not equal.", __PRETTY_FUNCTION__);
        return NO;
    }

    if (((toFrame + length) > self.frameLength) || ((fromFrame + length) > fromData.frameLength)) {
        YASLog(@"%s - Out of range (toFrame = %@ / fromFrame = %@ / length = %@ / toFrameLength = %@ / fromFrameLength "
               @"= %@).",
               __PRETTY_FUNCTION__, @(toFrame), @(fromFrame), @(length), @(self.frameLength), @(fromData.frameLength));
        return NO;
    }

    const UInt32 bytesPerFrame = self.format.streamDescription->mBytesPerFrame;

    for (UInt32 i = 0; i < self.bufferCount; i++) {
        Byte *toDataPointer = self.audioBufferList->mBuffers[i].mData;
        Byte *fromDataPointer = fromData.audioBufferList->mBuffers[i].mData;
        memcpy(&toDataPointer[toFrame * bytesPerFrame], &fromDataPointer[fromFrame * bytesPerFrame],
               length * bytesPerFrame);
    }

    return YES;
}

- (BOOL)copyFlexiblyFromData:(YASAudioData *)data
{
    if (!data) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Argument is nil. data(%@)", __PRETTY_FUNCTION__, data]));
        return NO;
    }

    if (self.format.bitDepthFormat != data.format.bitDepthFormat) {
        YASLog(@"%s - Invalid bit depth format.", __PRETTY_FUNCTION__);
        return NO;
    }

    const AudioBufferList *abl = data.audioBufferList;

    return [self copyFlexiblyFromAudioBufferList:abl];
}

- (BOOL)copyFlexiblyFromAudioBufferList:(const AudioBufferList *)fromAbl
{
    if (!fromAbl) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Argument is nil. fromAbl(%p)", __PRETTY_FUNCTION__, fromAbl]));
        return NO;
    }

    self.frameLength = 0;
    [self _resetDataByteSize];

    AudioBufferList *toAbl = self.mutableAudioBufferList;
    const UInt32 sampleByteCount = self.format.sampleByteCount;
    UInt32 frameLength = 0;

    BOOL result = YASAudioCopyAudioBufferListFlexibly(fromAbl, toAbl, sampleByteCount, &frameLength);

    if (result) {
        self.frameLength = frameLength;
    }

    return result;
}

- (BOOL)copyFlexiblyToAudioBufferList:(AudioBufferList *)toAbl
{
    if (!toAbl) {
        YASRaiseWithReason(([NSString stringWithFormat:@"%s - Argument is nil. toAbl(%p)", __PRETTY_FUNCTION__, toAbl]));
        return NO;
    }

    const AudioBufferList *fromAbl = self.audioBufferList;
    const UInt32 sampleByteCount = self.format.sampleByteCount;

    return YASAudioCopyAudioBufferListFlexibly(fromAbl, toAbl, sampleByteCount, NULL);
}

#pragma mark Private

- (void)_resetDataByteSize
{
    const AudioStreamBasicDescription *asbd = self.format.streamDescription;
    UInt32 dataByteSize = (UInt32)(self.frameCapacity * asbd->mBytesPerFrame);
    AudioBufferList *abl = self.mutableAudioBufferList;
    for (NSInteger i = 0; i < abl->mNumberBuffers; i++) {
        abl->mBuffers[i].mDataByteSize = dataByteSize;
    }
}

@end

File: YASAudio/Basic/yas_audio_data.cpp
//
//  yas_audio_data.cpp
//  Copyright (c) 2015 Yuki Yasoshima.
//

#include "yas_audio_data.h"
#include "yas_audio_format.h"
#include "yas_audio_channel_route.h"
#include "YASAudioUtility.h"
#include <string>
#include <exception>
#include <functional>
#include <iostream>

using namespace yas;

#pragma mark - private

class audio_data::impl
{
   public:
    const audio_format_ptr format;
    const AudioBufferList *abl_ptr;
    const UInt32 frame_capacity;
    UInt32 frame_length;

    impl(const audio_format_ptr &format, AudioBufferList *ptr, const UInt32 frame_capacity)
        : format(format),
          abl_ptr(ptr),
          frame_capacity(frame_capacity),
          frame_length(frame_capacity),
          _abl(nullptr),
          _data(nullptr)
    {
    }

    impl(const audio_format_ptr &format, abl_unique_ptr &&abl, abl_data_unique_ptr &&data, const UInt32 frame_capacity)
        : format(format),
          frame_capacity(frame_capacity),
          frame_length(frame_capacity),
          abl_ptr(abl.get()),
          _abl(std::move(abl)),
          _data(std::move(data))
    {
    }

    impl(const audio_format_ptr &format, abl_unique_ptr &&abl, const UInt32 frame_capacity)
        : format(format),
          frame_capacity(frame_capacity),
          frame_length(frame_capacity),
          abl_ptr(abl.get()),
          _abl(std::move(abl)),
          _data(nullptr)
    {
    }

   private:
    const abl_unique_ptr _abl;
    const abl_data_unique_ptr _data;
};

std::pair<abl_unique_ptr, abl_data_unique_ptr> yas::allocate_audio_buffer_list(const UInt32 buffer_count,
                                                                           const UInt32 channels, const UInt32 size)
{
    abl_unique_ptr abl_ptr((AudioBufferList *)calloc(1, sizeof(AudioBufferList) + buffer_count * sizeof(AudioBuffer)),
                           [](AudioBufferList *abl) { free(abl); });

    abl_ptr->mNumberBuffers = buffer_count;
    auto data_ptr = std::make_unique<std::vector<std::vector<UInt8>>>();
    if (size > 0) {
        data_ptr->reserve(buffer_count);
    } else {
        data_ptr = nullptr;
    }

    for (UInt32 i = 0; i < buffer_count; ++i) {
        abl_ptr->mBuffers[i].mNumberChannels = channels;
        abl_ptr->mBuffers[i].mDataByteSize = size;
        if (size > 0) {
            data_ptr->push_back(std::vector<UInt8>(size));
            abl_ptr->mBuffers[i].mData = data_ptr->at(i).data();
        }
    }

    return std::make_pair(std::move(abl_ptr), std::move(data_ptr));
}

static void set_data_byte_size(audio_data &data, const UInt32 data_byte_size)
{
    AudioBufferList *abl = data.audio_buffer_list();
    for (UInt32 i = 0; i < abl->mNumberBuffers; i++) {
        abl->mBuffers[i].mDataByteSize = data_byte_size;
    }
}

static void reset_data_byte_size(audio_data &data)
{
    const UInt32 data_byte_size =
        (const UInt32)(data.frame_capacity() * data.format()->stream_description().mBytesPerFrame);
    set_data_byte_size(data, data_byte_size);
}

#pragma mark - public

audio_data_ptr audio_data::create(const audio_format_ptr &format, AudioBufferList *abl)
{
    return audio_data_ptr(new audio_data(format, abl));
}

audio_data_ptr audio_data::create(const audio_format_ptr &format, const UInt32 frame_capacity)
{
    return audio_data_ptr(new audio_data(format, frame_capacity));
}

audio_data_ptr audio_data::create(const audio_format_ptr &format, const audio_data &data,
                                  const std::vector<channel_route_ptr> &channel_routes, const bool is_output)
{
    return audio_data_ptr(new audio_data(format, data, channel_routes, is_output));
}

audio_data::audio_data(const audio_format_ptr &format, AudioBufferList *abl)
{
    if (!format || !abl) {
        throw std::invalid_argument(std::string(__PRETTY_FUNCTION__) + " : argument is null.");
    }

    _impl = std::make_unique<impl>(format, abl,
                                   abl->mBuffers[0].mDataByteSize / format->stream_description().mBytesPerFrame);
}

audio_data::audio_data(const audio_format_ptr &format, const UInt32 frame_capacity)
{
    if (!format || frame_capacity == 0) {
        throw std::invalid_argument(std::string(__PRETTY_FUNCTION__) + " : argument is null.");
    }

    auto pair = allocate_audio_buffer_list(format->buffer_count(), format->stride(),
                                           frame_capacity * format->stream_description().mBytesPerFrame);
    _impl = std::make_unique<impl>(format, std::move(pair.first), std::move(pair.second), frame_capacity);
}

audio_data::audio_data(const audio_format_ptr &format, const audio_data &data,
                       const std::vector<channel_route_ptr> channel_routes, const bool is_output)
{
    if (!format || channel_routes.size() == 0) {
        throw std::invalid_argument(std::string(__PRETTY_FUNCTION__) + " : argument is null.");
    }

    if (format->channel_count() != channel_routes.size() || format->is_interleaved()) {
        throw std::invalid_argument(std::string(__PRETTY_FUNCTION__) + " : invalid format. format.channel_count(" +
                                    std::to_string(format->channel_count()) + ") channel_routes.size(" +
                                    std::to_string(channel_routes.size()) + ") is_interleaved(" +
                                    std::to_string(format->is_interleaved()) + ")");
    }

    auto pair = allocate_audio_buffer_list(format->buffer_count(), format->stride(), 0);
    abl_unique_ptr &to_abl = pair.first;

    const AudioBufferList *from_abl = data.audio_buffer_list();
    UInt32 bytesPerFrame = format->stream_description().mBytesPerFrame;
    UInt32 frame_capacity = 0;

    for (UInt32 i = 0; i < format->channel_count(); i++) {
        const channel_route_ptr &route = channel_routes.at(i);
        UInt32 from_channel = is_output ? route->destination_channel() : route->source_channel();
        UInt32 to_channel = is_output ? route->source_channel() : route->destination_channel();
        to_abl->mBuffers[to_channel].mData = from_abl->mBuffers[from_channel].mData;
        to_abl->mBuffers[to_channel].mDataByteSize = from_abl->mBuffers[from_channel].mDataByteSize;
        UInt32 frame_length = from_abl->mBuffers[0].mDataByteSize / bytesPerFrame;
        if (frame_capacity == 0) {
            frame_capacity = frame_length;
        } else if (frame_capacity != frame_length) {
            throw std::invalid_argument(std::string(__PRETTY_FUNCTION__) + " : invalid frame length. frame_capacity(" +
                                        std::to_string(frame_capacity) + ") frame_length(" +
                                        std::to_string(frame_length) + ")");
        }
    }

    _impl = std::make_unique<impl>(format, std::move(to_abl), frame_capacity);
}

audio_format_ptr audio_data::format() const
{
    return _impl->format;
}

AudioBufferList *audio_data::audio_buffer_list()
{
    return const_cast<AudioBufferList *>(_impl->abl_ptr);
}

const AudioBufferList *audio_data::audio_buffer_list() const
{
    return _impl->abl_ptr;
}

audio_pointer audio_data::audio_ptr_at_buffer(const UInt32 buffer) const
{
    if (buffer >= _impl->abl_ptr->mNumberBuffers) {
        throw std::out_of_range(std::string(__PRETTY_FUNCTION__) + " : out of range. buffer(" + std::to_string(buffer) +
                                ") _impl->abl_ptr.mNumberBuffers(" + std::to_string(_impl->abl_ptr->mNumberBuffers) +
                                ")");
    }

    return audio_pointer{.v = _impl->abl_ptr->mBuffers[buffer].mData};
}

audio_pointer audio_data::audio_ptr_at_channel(const UInt32 channel) const
{
    audio_pointer pointer{.v = nullptr};

    const UInt32 stride = format()->stride();
    const AudioBufferList *abl_ptr = _impl->abl_ptr;

    if (stride > 1) {
        if (channel < abl_ptr->mBuffers[0].mNumberChannels) {
            pointer.v = abl_ptr->mBuffers[0].mData;
            if (channel > 0) {
                pointer.u8 += channel * format()->sample_byte_count();
            }
        } else {
            throw std::out_of_range(std::string(__PRETTY_FUNCTION__) + " : out of range. channel(" +
                                    std::to_string(channel) + ") mNumberChannels(" +
                                    std::to_string(abl_ptr->mBuffers[0].mNumberChannels) + ")");
        }
    } else {
        if (channel < abl_ptr->mNumberBuffers) {
            pointer.v = abl_ptr->mBuffers[channel].mData;
        } else {
            throw std::out_of_range(std::string(__PRETTY_FUNCTION__) + " : out of range. channel(" +
                                    std::to_string(channel) + ") mNumberChannels(" +
                                    std::to_string(abl_ptr->mBuffers[0].mNumberChannels) + ")");
        }
    }

    return pointer;
}

const UInt32 audio_data::frame_capacity() const
{
    return _impl->frame_capacity;
}

const UInt32 audio_data::frame_length() const
{
    return _impl->frame_length;
}

void audio_data::set_frame_length(const UInt32 length)
{
    if (length > frame_capacity()) {
        throw std::out_of_range(std::string(__PRETTY_FUNCTION__) + " : out of range. frame_length(" +
                                std::to_string(length) + ") frame_capacity(" + std::to_string(frame_capacity()) + ")");
        return;
    }

    _impl->frame_length = length;

    const UInt32 data_byte_size = format()->stream_description().mBytesPerFrame * length;
    set_data_byte_size(*this, data_byte_size);
}

void audio_data::clear()
{
    set_frame_length(frame_capacity());
    yas::clear(audio_buffer_list());
}

void audio_data::clear(const UInt32 start_frame, const UInt32 length)
{
    if ((start_frame + length) > frame_length()) {
        throw std::out_of_range(std::string(__PRETTY_FUNCTION__) + " : out of range. frame(" +
                                std::to_string(start_frame) + " length(" + std::to_string(length) + " frame_length(" +
                                std::to_string(frame_length()) + ")");
    }

    const UInt32 bytes_per_frame = format()->stream_description().mBytesPerFrame;
    for (UInt32 i = 0; i < format()->buffer_count(); i++) {
        UInt8 *byte_data = static_cast<UInt8 *>(audio_buffer_list()->mBuffers[i].mData);
        memset(&byte_data[start_frame * bytes_per_frame], 0, length * bytes_per_frame);
    }
}

#pragma mark - global

void yas::clear(AudioBufferList *abl)
{
    for (UInt32 i = 0; i < abl->mNumberBuffers; ++i) {
        if (abl->mBuffers[i].mData) {
            memset(abl->mBuffers[i].mData, 0, abl->mBuffers[i].mDataByteSize);
        }
    }
}

audio_data::copy_result yas::copy_data(const audio_data_ptr &from_data, audio_data_ptr &to_data)
{
    return copy_data(from_data, to_data, 0, 0, from_data->frame_length());
}

audio_data::copy_result yas::copy_data(const audio_data_ptr &from_data, audio_data_ptr &to_data,
                                       const UInt32 from_start_frame, const UInt32 to_start_frame, const UInt32 length)
{
    if (!from_data || !to_data) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_argument);
    }

    if (*from_data->format() != *to_data->format()) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_format);
    }

    if (((to_start_frame + length) > to_data->frame_length()) ||
        ((from_start_frame + length) > from_data->frame_length())) {
        return audio_data::copy_result(audio_data::copy_error_type::out_of_range_frame_length);
    }

    const UInt32 bytes_per_frame = to_data->format()->stream_description().mBytesPerFrame;
    const UInt32 buffer_count = to_data->format()->buffer_count();

    for (UInt32 i = 0; i < buffer_count; i++) {
        UInt8 *to_data_ptr = static_cast<UInt8 *>(to_data->audio_buffer_list()->mBuffers[i].mData);
        const UInt8 *from_data_ptr = static_cast<const UInt8 *>(from_data->audio_buffer_list()->mBuffers[i].mData);
        memcpy(&to_data_ptr[to_start_frame * bytes_per_frame], &from_data_ptr[from_start_frame * bytes_per_frame],
               length * bytes_per_frame);
    }

    return audio_data::copy_result(nullptr);
}

audio_data::copy_result yas::copy_data_flexibly(const AudioBufferList *&from_abl, AudioBufferList *&to_abl,
                                                const UInt32 sample_byte_count, UInt32 *out_frame_length)
{
    if (YASAudioCopyAudioBufferListFlexibly(from_abl, to_abl, sample_byte_count, out_frame_length)) {
        return audio_data::copy_result(nullptr);
    } else {
        return audio_data::copy_result(audio_data::copy_error_type::flexible_copy_failed);
    }
}

audio_data::copy_result yas::copy_data_flexibly(const audio_data_ptr &from_data, audio_data_ptr &to_data)
{
    if (!from_data || !to_data) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_argument);
    }

    if (from_data->format()->pcm_format() != to_data->format()->pcm_format()) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_pcm_format);
    }

    const auto abl = from_data->audio_buffer_list();
    return copy_data_flexibly(abl, to_data);
}

audio_data::copy_result yas::copy_data_flexibly(const audio_data_ptr &from_data, AudioBufferList *to_abl)
{
    if (!from_data || !to_abl) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_argument);
    }

    const AudioBufferList *from_abl = from_data->audio_buffer_list();
    const UInt32 sample_byte_count = from_data->format()->sample_byte_count();
    return copy_data_flexibly(from_abl, to_abl, sample_byte_count, nullptr);
}

audio_data::copy_result yas::copy_data_flexibly(const AudioBufferList *from_abl, audio_data_ptr &to_data)
{
    if (!from_abl || !to_data) {
        return audio_data::copy_result(audio_data::copy_error_type::invalid_argument);
    }

    to_data->set_frame_length(0);
    reset_data_byte_size(*to_data);

    AudioBufferList *to_abl = to_data->audio_buffer_list();
    const UInt32 sample_byte_count = to_data->format()->sample_byte_count();
    UInt32 frameLength = 0;

    auto result = copy_data_flexibly(from_abl, to_abl, sample_byte_count, &frameLength);
    if (result) {
        to_data->set_frame_length(frameLength);
    }
    return result;
}

UInt32 yas::frame_length(const AudioBufferList *abl, const UInt32 sample_byte_count)
{
    if (sample_byte_count > 0) {
        UInt32 out_frame_length = 0;
        for (UInt32 buf = 0; buf < abl->mNumberBuffers; buf++) {
            const AudioBuffer *ab = &abl->mBuffers[buf];
            const UInt32 stride = ab->mNumberChannels;
            const UInt32 frame_length = ab->mDataByteSize / stride / sample_byte_count;
            if (buf == 0) {
                out_frame_length = frame_length;
            } else if (out_frame_length != frame_length) {
                return 0;
            }
        }
        return out_frame_length;
    } else {
        return 0;
    }
}
